File: leetcode/java/0206. Reverse Linked List.java
package leetcode.java;

// Time complexity: O(N)
// Space complexity: O(1)

// Provided class
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}


class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode res = null;
        ListNode curr = head;
        while (head != null) {
            ListNode temp = curr;
            head = head.next;
            curr = curr.next;
            temp.next = res;
            res = temp;
        }
        return res;
    }
}

File: leetcode/python/0206. Reverse Linked List.py
# Time complexity: O(N)
# Space complexity: O(1)

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def reverseList(self, head):
        prev = None
        curr = head
        while head != None:
            head = head.next
            curr.next = prev
            prev = curr
            curr = head
        return prev  